hnn_core/lfp.py

```diff
@@ -0,0 +1,433 @@
+    mask = np.logical_and(times >= plot_tmin, times < plot_tmax)
+    times = times[mask]
+
+    data = np.array(data)[mask]
+
+    return data, times
+
+
+class LFPArray:
+    """Local field potential (LFP) electrode array class.
+
+    Parameters
+    ----------
+    positions : tuple | list of tuple
+        The (x, y, z) coordinates (in um) of the LFP electrodes.
+    sigma : float
+        Extracellular conductivity, in S/m, of the assumed infinite,
+        homogeneous volume conductor that the cell and electrode are in.
+    method : str
+        Approximation to use. 'psa' (default) assigns all transmembrane
+        currents to the center point (0.5) (point source approximation).
+        'lsa' treats the section as a line source, and a single multiplier is
+        calculated for each section (line source approximation).
+
+    Attributes
+    ----------
+    voltages : list of float
+        The LFP voltages at the electrode positions (in uV).
+    times : list of float
+        The time points the LFP is sampled at (ms)
+    sfreq : float
+        Sampling rate of the LFP data (Hz).
+
+    Notes
+    -----
+    See Table 5 in http://jn.physiology.org/content/104/6/3388.long for
+    measured values of sigma in rat cortex (note units there are mS/cm)
+    """
+
+    def __init__(self, positions, sigma=0.3, method='psa', times=list(),
+                 voltages=list()):
+        self.positions = positions
+        self.sigma = sigma
+        self.method = method
+        self.times = times
+        self._data = voltages
+
+    def __getitem__(self, trial_no):
+        if isinstance(trial_no, int):
+            return_data = [self._data[trial_no]]
+        elif isinstance(trial_no, slice):
+            return_data = self._data[trial_no]
+        elif not isinstance(trial_no, (list, tuple)):
+            return_data = [self._data[trial] for trial in trial_no]
+
+        return LFPArray(self.positions, sigma=self.sigma, method=self.method,
+                        times=self.times, voltages=return_data)
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        msg = (f'{len(self.positions)} electrodes, sigma={self.sigma}, '
+               f'method={self.method}')
+        if len(self._data) > 0:
+            msg += f' | {len(self._data)} trials, {len(self.times)} times'
+        return f'<{class_name} | {msg}>'
+
+    def __len__(self):
+        return len(self.positions)
+
+    @property
+    def sfreq(self):
+        """Return the sampling rate of the LFP data."""
+        if len(self.times) > 1:
+            return 1000. / (self.times[1] - self.times[0])  # times in ms
+        else:
+            return None
+
+    def reset(self):
+        self._data = list()
+        self.times = list()
+
+    def get_data(self):
+        return np.array(self._data)
+
+    def plot(self, *, trial_no=None, contact_no=None, window_len=None,
+             tmin=None, tmax=None, ax=None, decim=None, color=None, show=True):
```

Comment:
Okay I didn't notice this. Do we still need to expose `trial_no` and `contact_no`? If `__getitem__` returns an instance of `LFPArray`, then you should be able to do `net.lfp['shank1'][trial_no].plot()` without having to specify `trial_no` and `contact_no`, no?

Reply:
Nope :( The `plot`-method needs to be able to handle multiple trials, since `__getitem__` can be a slice. The `_data`-attribute is always 3D `[trials, contacts, times]`, and there has to be a way to specify a subset of contacts to plot, too. So though both `trial_no` and `contact_no` are optional (default: `None`==plot everything on top of each other), they need to be there, methinks.

Reply:
humm ... but you have to be very careful with the indexing if you supply a slice (e.g., `net.lfp['shank1'][2:5]].plot(trialno=2)` and then do `trial_no` in the `plot` method, it can easily be screwed up. Makes me think different trials should be different subplots. Or if there are too many trials, then it should actually be some kind of `matshow` plot.